API documentation

mattias-p opened the pull request:
This integrates notes_api.md into API.md, adds lots of details, and splits it all up into a number of different documents and links to all of them from the documentation section of README.

I'd say this is an improvement over the previous situation, but it's far from complete. For instance there are still almost 50 "TODO" notes in API.md.

matsduf commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+* `"ds_info"`: A list of *DS info* objects, optional.
+* `"advanced"`: A boolean, optional.
+* `"ipv6"`: A boolean, optional.
+* `"ipv4"`: A boolean, optional.
```
Add that default of both ipv4 and ipv6 is "false".

matsduf commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+* `"id"`: A *test id*, required.
+* `"language"`: A string, required. Must be at least two characters long. The
+  two first characters are used to look up the *translation language* to be
+  used. If the lookup fails, the choice defaults to English.
```
List supported language tags. If the lookup the messages can be broken since the even for English "translation" is needed.

mattias-p replied:
The set of translation languages is configurable, right? And it's the Zonemaster Engine configuration that's the authority? I think the Backend documentation should link into the Engine documentation where it describes how you can inspect the set of translation languages.

matsduf replied:
No, it is hardcoded in the backend as far as I can see:

```
    if ( $browser_lang eq 'fr' ) {
        setlocale( LC_ALL, "fr_FR.UTF-8" );
    }
    elsif ( $browser_lang eq 'sv' ) {
        setlocale( LC_ALL, "sv_SE.UTF-8" );
    }
    else {
        setlocale( LC_ALL, "en_US.UTF-8" );
    }
```

From Zonemaster/WebBackend/Translator.pm

matsduf commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+
+* `"offset"`: An integer, optional. (default: 0).
+* `"limit"`: An integer, optional. (default: 200).
+* `"frontend_params"`: As described below.
```
Is "offset" and "limit" really implemented?

mattias-p replied:
This is from the current documentation. If it's not implemented that's a bug, right?

matsduf replied:
@mtoma, what is the status of "offset" and "limit"? I can see code for Mysql and PostgreSQL, but is it working?

mattias-p replied:
I can limit results by twiddling with the "offset" and "limit" parameters.

matsduf commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+
+If an identical *test* for a domain was already enqueued and hasn't been started or was enqueued less than 10 minutes earlier,
+no new *test* is enqueued for this domain.
+
```
A "batch user" cannot start a new batch before the first batch has completed. If a second "batch user" is created, then a second batch can be started in parallel with the first batch.

mattias-p replied:
What's the behavior when an api user tries to add a second batch while his first batch is running? I haven't succeeded in adding any api users, and I haven't had time to investigate it further, so I can't check this myself right now.

matsduf replied:
You get an error back saying that you already run a batch:

```json
{
  "error": {
    "message": "You can't create a new batch job, job:[7] started on:[2016-11-09 10:39:02] still running  at /usr/local/share/perl/5.18.2/Zonemaster/WebBackend/DB/MySQL.pm line 89.\n",
    "code": -32603
  },
  "id": "MD4712",
  "jsonrpc": "2.0"
}
```

mattias-p replied:
Documented this as an error state.

matsduf commented on docs/Architecture.md:
```diff
@@ -0,0 +1,87 @@
+to the *Database*. A single *Worker* may handle several requests concurrently.
+The *Backend* architecture supports a single *Workers* interacting with a single
+*Database*.
+
```
If the workers picks the tests from different queues.

matsduf replied:
`single *Workers* interacting` --> `single *Worker* interacting`

mattias-p commented on README.md:
```diff
@@ -35,14 +35,26 @@ instructions](docs/installation.md).
 
```
Isn't this section supposed to be about listing and documenting the individual configuration options rather than the simplest possible steps to the application up and running (which is what's described in the installation guide)?

matsduf replied:
My comment refer to line 33, which is one line before what we see. It was not clear.

matsduf commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+   if needed), each component part must be at most 63 characters long.
+
+5. Each such component part must also consist only of the characters
+   `0` to `9`, `a` to `z`, `A` to `Z` and `-`.
```
If we have the classical limitation (as BASIC00 requires) then the label may not start nor end with a "-".

But as it is implemented, there are no visible limitations on the character set. The domain can contains any character. This is in conflict with BASIC00.

mattias-p replied:
I understand it that BASIC00 is the matter of some controversy at the moment, is that correct? If so, shouldn't we defer things that depend on it until its issues have been straightened out? I figure we'll need to go over things after the specification is settled anyhow.

matsduf replied:
* Bullet 1 gives a correct picture of the behavior of the backend. It does impose this features, but too generously (even invalid U-labels are converted to A-labels).
* Bullet 2-4 gives a correct picture of the behavior of the backend. It does impose those restrictions.
* Bullet 5 does not give a correct picture of the behavior of the backend. It does not impose those restrictions.

BASIC00 is a different, even though related, question.

mattias-p replied:
So I will remove item 5 as this requirement is not imposed.

What should I do about item 1? Remove it? Update its wording? To what?

matsduf replied:
"Currently there are no restrictions on what characters that are allowed."

mattias-p commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+### Deviations from JSON-RPC 2.0
+
+* The `"jsonrpc"` property is not checked.
+* The error code -32603 is used for invalid arguments, as opposed to -32602.
```
Fixing with a slightly different wording.

mattias-p commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+
+* Extra top-level properties in request objects are allowed but ignored.
+* Extra properties in the `"params"` object are allowed for some methods but ignored for others.
+* Error messages from the API should be considered sensitive sas they leak details about the internals of the application and the system.
```
Fixing with a slightly different wording.

mattias-p commented on docs/API.md:
```diff
@@ -1,681 +1,1085 @@
+* Extra top-level properties in request objects are allowed but ignored.
+* Extra properties in the `"params"` object are allowed for some methods but ignored for others.
+* Error messages from the API should be considered sensitive sas they leak details about the internals of the application and the system.
+* The error code -32601 is used when the `"method"` property is missing, as opposed to -32600.
```
Fixing with a slightly different wording.

mtoma commented:
It is implemented but not tested. Will require under 1h of work to fix if necessary.

----- Original Message -----
From: "Mattias Päivärinta" <notifications@github.com>
To: "dotse/zonemaster-backend" <zonemaster-backend@noreply.github.com>
Sent: Friday, November 25, 2016 1:47:17 PM
Subject: Re: [dotse/zonemaster-backend] API documentation (#233)

mattias-p commented on this pull request.



>  
-**Request**:
-```
+An object with the following properties:
+
+* `"offset"`: An integer, optional. (default: 0).
+* `"limit"`: An integer, optional. (default: 200).
+* `"frontend_params"`: As described below.

This is from the current documentation. If it's not implemented that's a bug, right?

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/dotse/zonemaster-backend/pull/233

sandoche2k commented on README.md:
```diff
@@ -35,14 +35,26 @@ instructions](docs/installation.md).
-There is a fully documented [API](docs/API.md), which is the primay way
-to use the backend. The [docs](docs/) directory also contains the SQL commands
-for manipulating the database. 
+The Zonemaster Backend documentation is split up into several documents:
```
Test

mtoma commented:
The API behavior when it finds one or more tests with progress < 100 for a batch job from a user that tries to add another batch job is to refuse to add the new batch job.

----- Original Message -----
From: "Mattias Päivärinta" <notifications@github.com>
To: "dotse/zonemaster-backend" <zonemaster-backend@noreply.github.com>
Sent: Friday, November 25, 2016 1:52:49 PM
Subject: Re: [dotse/zonemaster-backend] API documentation (#233)

mattias-p commented on this pull request.



> +>
+> TODO: List all possible error codes and describe what they mean enough for clients to know how react to them.
+>
+
+
+## API method: `add_batch_job`
+
+>
+> TODO: Method description.
+>
+
+All the domains will be tested using identical parameters.
+
+If an identical *test* for a domain was already enqueued and hasn't been started or was enqueued less than 10 minutes earlier,
+no new *test* is enqueued for this domain.
+

What's the behavior when an api user tries to add a second batch while his first batch is running? I haven't succeeded in adding any api users, and I haven't had time to investigate it further, so I can't check this myself right now.

-- 
You are receiving this because you are subscribed to this thread.
Reply to this email directly or view it on GitHub:
https://github.com/dotse/zonemaster-backend/pull/233

matsduf commented on docs/Architecture.md:
```diff
@@ -0,0 +1,87 @@
+keeping records of performed domain health checks.
+
+A Zonemaster *Backend* system consists of at least three components: a
+single *Database*, one or more *Workers* and one or more *Web backends*.
```
Only one Worker unless we also document the queues.

mattias-p commented:
This PR resolves some but not all items in #176.
 * API and Notes_API have been fully merged.
 * Placeholders for missing information are very prevalent, but probably not complete.
 * The request format is explained to some degree, but it is not fully specified.
 * The response format is mostly descibed by example and still needs much improvement.
 * Error responses are hardly described at all.

In my view:
 * I have addressed all issues that came up during the review of this PR.
 * There is consensus within the team that the current state of this PR is an improvement over the current state of the master branch.

If nobody disagrees with the above, this can be merged.

matsduf commented:
Does anyone have any problem if I merge this?
